ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -929,90 +942,185 @@ private void updateClusterMapInHelix(boolean startValidatingClusterManager) thro
+              --totalInstances);
+          if (helixAdminOperation == HelixAdminOperation.BootstrapCluster) {
+            if (!dryRun) {
+              removeDataNodeConfigFromHelix(dcName, instanceName, propertyStoreAdapter);
+            }
+            instancesDropped.getAndIncrement();
           }
-          instancesDropped.getAndIncrement();
+        } else {
+          info(
+              "[{}] Instance {} is in Helix {}, but not in static. Ignoring for now (use --forceRemove to forcefully remove). "
+                  + "Remaining instances: {}", dcName.toUpperCase(), instanceName, dataNodeConfigSourceType.name(),
+              --totalInstances);
+          expectMoreInHelixDuringValidate = true;
+          instancesNotForceRemovedByDc.computeIfAbsent(dcName, k -> ConcurrentHashMap.newKeySet()).add(instanceName);
         }
-      } else {
-        info(
-            "[{}] Instance {} is in Helix, but not in static. Ignoring for now (use --forceRemove to forcefully remove). "
-                + "Remaining instances: {}", dcName.toUpperCase(), instanceName, --totalInstances);
-        expectMoreInHelixDuringValidate = true;
-        instancesNotForceRemovedByDc.computeIfAbsent(dcName, k -> ConcurrentHashMap.newKeySet()).add(instanceName);
       }
     }
   }
 
+  private List<String> getInstanceNamesInHelix(String dcName, PropertyStoreToDataNodeConfigAdapter adapter) {
+    List<String> instanceNames;
+    if (dataNodeConfigSourceType == DataNodeConfigSourceType.PROPERTY_STORE) {
+      instanceNames = adapter.getAllDataNodeNames();
+    } else {
+      instanceNames = adminForDc.get(dcName).getInstancesInCluster(clusterName);
+    }
+    return instanceNames;
+  }
+
+  private DataNodeConfig createDataNodeConfigFromStatic(String dcName, String instanceName,
+      DataNodeConfig referenceConfig, Map<String, Set<String>> partitionsToInstancesInDc,
+      InstanceConfigToDataNodeConfigAdapter.Converter converter) {
+    InstanceConfig referenceInstanceConfig =
+        overrideReplicaStatus || referenceConfig == null ? null : converter.convert(referenceConfig);
+    return converter.convert(
+        createInstanceConfigFromStaticInfo(dcToInstanceNameToDataNodeId.get(dcName).get(instanceName),
+            partitionsToInstancesInDc, instanceToDiskReplicasMap, referenceInstanceConfig));
+  }
+
+  private DataNodeConfig getDataNodeConfigFromHelix(String dcName, String instanceName,
+      PropertyStoreToDataNodeConfigAdapter adapter, InstanceConfigToDataNodeConfigAdapter.Converter converter) {
+    DataNodeConfig dataNodeConfig;
+    if (dataNodeConfigSourceType == DataNodeConfigSourceType.PROPERTY_STORE) {
+      dataNodeConfig = adapter.get(instanceName);
+    } else {
+      dataNodeConfig = converter.convert(adminForDc.get(dcName).getInstanceConfig(clusterName, instanceName));
+    }
+    return dataNodeConfig;
+  }
+
+  private boolean areDataNodeConfigsEquivalent(DataNodeConfig configInStatic, DataNodeConfig configInHelix,
```

Comment:
Maybe we could move this method into `DataNodeConfig`

`public boolean equals(DataNodeConfig other, boolean ignorePartitionStateConfigs)` (or whatever name you feel is best for the sealed/stopped/disabled fields)

Reply:
Sure, it makes more sense.